File: partiql-spi/src/main/java/org/partiql/spi/datetime/Timestamp.java
package org.partiql.spi.datetime;


import org.jetbrains.annotations.NotNull;

import java.time.LocalDateTime;

/**
 * A PartiQL timestamp without timezone.
 */
public class Timestamp {

    private final Date date;
    private final Time time;

    /**
     * NO PUBLIC CONSTRUCTORS.
     */
    private Timestamp(@NotNull Date date, @NotNull Time time) {
        this.date = date;
        this.time = time;
    }

    /**
     * @return timestamp for current date-time for the system clock.
     */
    @NotNull
    public Timestamp now() {
        return new Timestamp(Date.now(), Time.now());
    }

    /**
     * Create a new Timestamp from the date-time pair.
     *
     * @param date a date.
     * @param time a time.
     * @return new Timestamp
     */
    @NotNull
    public static Timestamp of(@NotNull Date date, @NotNull Time time) {
        return new Timestamp(date, time);
    }

    /**
     * Create a new Timestamp from the fields.
     *
     * @param year   the year to represent, from Year.MIN_YEAR to Year.MAX_YEAR
     * @param month  the month-of-year to represent, from 1 (January) to 12 (December)
     * @param day    the day of month to represent, from 1 to 31
     * @param hour   hour-of-day from 0 to 23
     * @param minute minute-of-hour from 0 to 59
     * @param second second-of-minute from 0 to 59
     * @return new Timestamp
     */
    @NotNull
    public static Timestamp of(int year, int month, int day, int hour, int minute, int second) {
        Date date = Date.of(year, month, day);
        Time time = Time.of(hour, minute, second);
        return new Timestamp(date, time);
    }

    /**
     * @param year   the year to represent, from Year.MIN_YEAR to Year.MAX_YEAR
     * @param month  the month-of-year to represent, from 1 (January) to 12 (December)
     * @param day    the day of month to represent, from 1 to 31
     * @param hour   hour-of-day from 0 to 23
     * @param minute minute-of-hour from 0 to 59
     * @param second second-of-minute from 0 to 59
     * @param nano   nano-of-second from 0 to 999,999,999
     * @return new Timestamp
     */
    @NotNull
    public static Timestamp of(int year, int month, int day, int hour, int minute, int second, int nano) {
        Date date = Date.of(year, month, day);
        Time time = Time.of(hour, minute, second, nano);
        return new Timestamp(date, time);
    }

    /**
     * Create a Timestamp from a {@link LocalDateTime}.
     *
     * @param timestamp local date time (without timezone).
     * @return new Timestamp
     */
    @NotNull
    public static Timestamp of(@NotNull LocalDateTime timestamp) {
        return new Timestamp(Date.of(timestamp.toLocalDate()), Time.of(timestamp.toLocalTime()));
    }

    /**
     * @return a {@link LocalDateTime} for this timestamp.
     */
    @NotNull
    public LocalDateTime toLocalDateTime() {
        return LocalDateTime.of(date.toLocalDate(), time.toLocalTime());
    }

    /**
     * @return the timestamp date.
     */
    @NotNull
    public Date toDate() {
        return date;
    }

    /**
     * @return the timestamp time.
     */
    @NotNull
    public Time toTime() {
        return time;
    }

    /**
     * @return the year of this date
     */
    public int getYear() {
        return date.getYear();
    }

    /**
     * @return the month (of the year) field.
     */
    public int getMonth() {
        return date.getMonth();
    }

    /**
     * @return the day-of-month field.
     */
    public int getDay() {
        return date.getDay();
    }

    /**
     * @return the hour-of-day, from 0 to 23
     */
    public int getHour() {
        return time.getHour();
    }

    /**
     * @return the minute-of-hour, from 0 to 59
     */
    public int getMinute() {
        return time.getMinute();
    }

    /**
     * @return the second-of-minute, from 0 to 59
     */
    public int getSecond() {
        return time.getSecond();
    }

    /**
     * @return the nano-of-second, from 0 to 999,999,999
     */
    public int getNano() {
        return time.getNano();
    }

    @Override
    public int hashCode() {
        return toLocalDateTime().hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null) return false;
        if (!(obj instanceof Timestamp)) return false;
        Timestamp o = (Timestamp) obj;
        return time.equals(o.time) && date.equals(o.date);
    }

    @Override
    public String toString() {
        // TODO confirm SQL format
        return "TIMESTAMP " + date + " " + time;
    }
}

File: partiql-spi/src/main/kotlin/org/partiql/spi/datetime/Timestamp.kt
package org.partiql.spi.datetime

import org.partiql.spi.datetime.util.DatetimeUtil.toBigDecimal
import org.partiql.spi.datetime.impl.OffsetTimestampLowPrecision
import org.partiql.spi.datetime.util.DatetimeComparisons
import org.partiql.spi.datetime.util.DatetimePrecisionUtil
import java.math.BigDecimal

/**
 * Superclass for all implementations representing time value.
 * See [TimestampWithTimeZone] and [TimestampWithoutTimeZone]
 */
public sealed interface Timestamp : Datetime, Comparable<Timestamp> {

    public override val year: Int
    public override val month: Int
    public override val day: Int
    public override val hour: Int
    public override val minute: Int
    public override val decimalSecond: BigDecimal
    public override val timeZone: Timezone?

    // Operation
    /**
     * Returns a [Timestamp] value with the specified number of years added.
     * [years] can be negative.
     */
    public fun plusYears(years: Long): Timestamp

    /**
     * Returns a [Timestamp] value with the specified number of years added.
     * Year fields may be changed to ensure the correctness of the result.
     * [months] can be negative.
     */
    public fun plusMonths(months: Long): Timestamp

    /**
     * Returns a [Timestamp] value with the specified number of days added.
     * Year, Month fields may be changed to ensure the correctness of the result.
     * [days] can be negative.
     */
    public fun plusDays(days: Long): Timestamp

    /**
     * Returns a [Timestamp] value with the specified number of hours added.
     * Year, Month, Day fields may be changed to ensure the correctness of the result.
     * [hours] can be negative.
     */
    public fun plusHours(hours: Long): Timestamp

    /**
     * Returns a [Timestamp] value with the specified number of minutes added.
     * Year, Month, Day, Hour fields may be changed to ensure the correctness of the result.
     * [minutes] can be negative.
     */
    public fun plusMinutes(minutes: Long): Timestamp

    /**
     * Returns a [Timestamp] value with the specified number of minutes added.
     * Year, Month, Day, Hour, Minute fields may be changed to ensure the correctness of the result.
     * [seconds] can be negative.
     */
    public fun plusSeconds(seconds: BigDecimal): Timestamp

    /**
     * Convenient wrapper for entering long value as [seconds]
     */
    public fun plusSeconds(seconds: Long): Timestamp

    /**
     * Returns a [Timestamp] value with [precision] number of digits in second fraction.
     */
    public fun toPrecision(precision: Int): Timestamp

    /**
     * Returns a [Date] value by extracting [year], [month], [day] field
     */
    public fun toDate(): Date = DateTimeValue.date(this.year, this.month, this.day)

    /**
     * Returns a [Time] value by extracting [hour], [minute], [decimalSecond], and [timeZone].
     */
    public fun toTime(): Time

    /**
     * Comparison method for timestamp value
     * If both value are [TimestampWithTimeZone]:
     *    The two value are considered equivalent if they refer to the same point in time.
     * If both value are timestamp without timezone:
     *      The two value are consider equivalent if all the fields (exclude precision) are equivalent.
     *      Another way to interpret this is, two timestamp without timezone are equivalent
     *          if and only if they are equivalent when converted to the same time zone.
     * If one value is timestamp with time zone and the other is timestamp without time zone:
     *      Error.
     *      One may not directly compare a timestamp value with timezone to a timestamp value without time zone.
     */
    public override fun compareTo(other: Timestamp): Int =
        DatetimeComparisons.compareTo(this, other)
}

/**
 * Superclass for all implementations representing timestamp with time zone value.
 * A [TimestampWithTimeZone] is a instant in the timeline.
 */
public abstract class TimestampWithTimeZone : Timestamp {
    public abstract override val timeZone: Timezone

    public abstract val ionRaw: com.amazon.ion.Timestamp?

    /**
     * Returns an Ion Equivalent Timestamp
     * [ionTimestampValue] takes care of the precision,
     * the ion representation will have exact precision of the backing partiQL value.
     * TODO: Move this to serde package.
     */
    @get:Throws(DateTimeException::class)
    public val ionTimestampValue: com.amazon.ion.Timestamp by lazy {
        when (val timeZone = this.timeZone) {
            Timezone.UnknownTimeZone -> com.amazon.ion.Timestamp.forSecond(
                this.year, this.month, this.day,
                this.hour, this.minute, this.decimalSecond,
                com.amazon.ion.Timestamp.UNKNOWN_OFFSET
            )

            is Timezone.UtcOffset -> com.amazon.ion.Timestamp.forSecond(
                this.year, this.month, this.day,
                this.hour, this.minute, this.decimalSecond,
                timeZone.totalOffsetMinutes
            )
        }
    }

    /**
     * Returns a BigDecimal representing the instant in time that is
     * the number of Seconds (*including* any fractional Seconds) from the epoch.
     * If the instant is before 1970-01-01T00:00:00Z, then the result will be negative.
     */
    public abstract val epochSecond: BigDecimal

    public val epochMillis: BigDecimal by lazy {
        epochSecond.movePointRight(3)
    }

    public abstract override fun plusYears(years: Long): TimestampWithTimeZone
    public abstract override fun plusMonths(months: Long): TimestampWithTimeZone
    public abstract override fun plusDays(days: Long): TimestampWithTimeZone
    public abstract override fun plusHours(hours: Long): TimestampWithTimeZone
    public abstract override fun plusMinutes(minutes: Long): TimestampWithTimeZone
    public abstract override fun plusSeconds(seconds: BigDecimal): TimestampWithTimeZone
    public final override fun plusSeconds(seconds: Long): TimestampWithTimeZone =
        plusSeconds(seconds.toBigDecimal())
    public final override fun toPrecision(precision: Int): TimestampWithTimeZone =
        DatetimePrecisionUtil.toPrecision(precision, this) as TimestampWithTimeZone

    /**
     * Returns a [TimestampWithoutTimeZone] object, assuming the local time is in [timeZone].
     */
    public abstract fun toTimeWithoutTimeZone(timeZone: Timezone): TimestampWithoutTimeZone

    /**
     * Returns a [TimestampWithTimeZone] object, assuming the desiredZone is [timeZone]
     * The intention of this function is to get the same instant of time in the desired zone.
     */
    public abstract fun atTimeZone(timeZone: Timezone): TimestampWithTimeZone
    public final override fun toTime(): TimeWithTimeZone =
        DateTimeValue.time(hour, minute, decimalSecond, timeZone) as TimeWithTimeZone

    public final override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other?.javaClass != javaClass) return false
        other as TimestampWithTimeZone
        if (this.year != other.year) return false
        if (this.month != other.month) return false
        if (this.day != other.day) return false
        if (this.hour != other.hour) return false
        if (this.minute != other.minute) return false
        if (this.decimalSecond != other.decimalSecond) return false
        if (this.timeZone != other.timeZone) return false
        return true
    }

    public final override fun hashCode(): Int =
        year.hashCode() + month.hashCode() + day.hashCode() + hour.hashCode() + minute.hashCode() + decimalSecond.hashCode() + timeZone.hashCode()

    public final override fun toString(): String =
        "${this.javaClass.simpleName}(year=$year, month=$month, day=$day, hour=$hour, minute=$minute, second=$decimalSecond, timeZone=$timeZone)"

    public companion object {

        @JvmStatic
        public fun nowZ(): TimestampWithTimeZone = OffsetTimestampLowPrecision.nowZ()
    }
}

/**
 * Superclass for all implementations representing timestamp without time zone value.
 * A [TimestampWithoutTimeZone] is not an instant in the timeline, the same value may refer to different instant in timeline at different time zone.
 */
public abstract class TimestampWithoutTimeZone : Timestamp {
    public override val timeZone: Timezone? = null
    public abstract override fun plusYears(years: Long): TimestampWithoutTimeZone
    public abstract override fun plusMonths(months: Long): TimestampWithoutTimeZone
    public abstract override fun plusDays(days: Long): TimestampWithoutTimeZone
    public abstract override fun plusHours(hours: Long): TimestampWithoutTimeZone
    public abstract override fun plusMinutes(minutes: Long): TimestampWithoutTimeZone
    public abstract override fun plusSeconds(seconds: BigDecimal): TimestampWithoutTimeZone
    public override fun plusSeconds(seconds: Long): TimestampWithoutTimeZone =
        plusSeconds(seconds.toBigDecimal())
    public final override fun toPrecision(precision: Int): TimestampWithoutTimeZone =
        DatetimePrecisionUtil.toPrecision(precision, this) as TimestampWithoutTimeZone

    /**
     * Returns a [TimestampWithTimeZone] object, assuming the current [TimestampWithoutTimeZone] value is in [timeZone].
     * The intention of this function is to append time zone information.
     */
    public abstract fun withTimeZone(timeZone: Timezone): TimestampWithTimeZone
    public final override fun toTime(): TimeWithoutTimeZone =
        DateTimeValue.time(this.hour, this.minute, this.decimalSecond) as TimeWithoutTimeZone
    public final override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other?.javaClass != javaClass) return false
        other as TimestampWithoutTimeZone
        if (this.year != other.year) return false
        if (this.month != other.month) return false
        if (this.day != other.day) return false
        if (this.hour != other.hour) return false
        if (this.minute != other.minute) return false
        if (this.decimalSecond != other.decimalSecond) return false
        return true
    }

    public final override fun hashCode(): Int =
        year.hashCode() + month.hashCode() + day.hashCode() + hour.hashCode() + minute.hashCode() + decimalSecond.hashCode()

    public final override fun toString(): String =
        "${this.javaClass.simpleName}(year=$year, month=$month, day=$day, hour=$hour, minute=$minute, second=$decimalSecond)"
}